src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,178 @@
+        return nodeByIndex.item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        Node<T> nodeByIndex = getNodeAndCheckIndex(index);
+        T oldValue = nodeByIndex.item;
+        nodeByIndex.item = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        Node<T> removedNote = getNodeAndCheckIndex(index);
+        unlink(removedNote, index);
+        size--;
+        return removedNote.item;
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> removedNote = head;
+        int counter = 0;
+        while (counter < size) {
+            if (removedNote.item == object
+                    || removedNote.item != null && removedNote.item.equals(object)) {
+                remove(counter);
+                return true;
+            }
+            removedNote = removedNote.next;
+            counter++;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void checkIndexOfNode(int index) {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException();
+        }
+    }
+
+    private static class Node<T> {
+        private T item;
+        private Node<T> next;
+        private Node<T> prev;
+
+        Node(Node<T> prev, T element, Node<T> next) {
+            this.item = element;
+            this.next = next;
+            this.prev = prev;
+        }
+    }
+
+    private Node<T> getNodeAndCheckIndex(int index) {
+        checkIndexOfNode(index);
+        Node<T> target = new Node<>(null,null,null);
+        int counter;
+        if (index <= (size / 2)) {
+            target = head;
+            counter = 0;
+            while (counter < index) {
+                target = target.next;
+                counter++;
+            }
+        } else if (index > (size / 2)) {
+            target = tail;
+            counter = size - 1;
+            while (counter > index) {
+                target = target.prev;
+                counter--;
+            }
+        }
+
+        return target;
+    }
+
+    private Node<T> getNode(int index) {
+        Node<T> target = new Node<>(null,null,null);
+        int counter;
+        if (index < (size / 2)) {
+            target = head;
+            counter = 0;
+            while (counter < index) {
+                target = target.next;
+                counter++;
+            }
+        } else if (index > (size / 2)) {
+            target = tail;
+            counter = size - 1;
+            while (counter > index) {
+                target = target.prev;
+                counter--;
+            }
+        }
+        return target;
+    }
+
+    private void unlink(Node<T> removedNote, int index) {
```

Comment:
```suggestion
    private void unlink(Node<T> removedNode) {
```
we should be able to unlink node without knowing its index
(you can check `removedNode.prev == null` for this)